src/core/base/model.storage.ts

```diff
@@ -0,0 +1,79 @@
+import type {UIPStateModel} from './model';
+
+interface UIPStateStorageEntry {
+  ts: string;
+  snippets: string;
+}
+
+interface UIPStateModelSnippets {
+  js: string;
+  html: string;
+  note: string;
+}
+
+export class UIPStateStorage {
+  public static readonly STORAGE_KEY = 'uip-editor-storage';
+
+  protected static readonly EXPIRATION_TIME = 3600000 * 12; // 12 hours
+
+  private static instances = new Map<string, UIPStateStorage>();
+
+  protected constructor(protected storeKey: string, protected model: UIPStateModel) {}
+
+  public static for(storeKey: string, model: UIPStateModel): UIPStateStorage {
```

Comment:
It seems to be we have cross reference and responsibilities.

Here is a couple of points architecturally lacks:

1) UIPRoot <1> --- <1> UIPStateModel

UIPRoot contains only possible storeKey (id) to save stat into the storage
Which means memoization with two criteria is redundant

2) The actual UIPStateStorage is stateless in reality - it does not own any data except store key which is just an action property

Points 1 + 2 means that memoization and singleton patterns are inapplicable here
Additionally point 1 gives us a clue that some of UIPRoot or UIPStateModel should be used as a holder in case we really want to have an actual UIPStateStorage object.
PS: UIPStateStorage object is not required at all (it is cosmetical and optional approach here) classically it seems to be an utility class

3) Point 3 I'd already mentioned in the UIPStateModel review
We introduce cycling direct references which means in 99% of cases bad architecture and mixed responsibilities.
If we go back to original UIP team idea: The UIPRoot is a HOLDER for  UIPStateModel 
That means UIPRoot contains and depends on UIPStateModel but not vice versa

4) From DX point of view - we want to have 2 actions  save state with a key, restore state.
That practically means one of the 2 possible architectures:
 I.  UIPRoot is the holder of save and restore logic.  UIPStateModel does not work with the store at all. Any sate observation to save current state works inderectly via observer pattern.
Once again UIPStateModel DOES not take any responsibility of that functionality at all in that case. We are free to create Delegate or Utility (called UIPStateStorage or whatever) to actually run this logic. UIPStateModel know nothing about UIPStateStorage.

 II. UIPStateStorage is a holder of  save and restore logic. Practically we should make store key as an own property of  UIPStateStorage (that's fine UIPRoot could initialize Model with that value, there is no any problems).  Both actions does not depend on something except the Model itself. We are technically free to choose of creation Utility or temporary instance class (again called UIPStateStorage or whatever) to implement logic in getails.
 
 in both cases core API should look like that:
 
 ```
   (root | model).reset(source?: 'html' | 'js'); 
   // save method could be presented but actually called by the inner implementation
 ```